enterprise/frontend/src/embedding-sdk/hooks/private/use-load-question.ts

```diff
@@ -0,0 +1,123 @@
+import { useRef, useState } from "react";
+import { useAsyncFn, useUnmount } from "react-use";
+
+import {
+  runQuestionOnLoadSdk,
+  runQuestionOnQueryChangeSdk,
+  runQuestionOnNavigateSdk,
+} from "embedding-sdk/lib/interactive-question";
+import type {
+  LoadSdkQuestionParams,
+  NavigateToNewCardParams,
+  SdkQuestionResult,
+} from "embedding-sdk/types/question";
+import { defer, type Deferred } from "metabase/lib/promise";
+import { useDispatch } from "metabase/lib/redux";
+import type Question from "metabase-lib/v1/Question";
+
+export interface LoadQuestionHookResult {
+  question?: Question;
+  queryResults?: any[];
+
+  isQuestionLoading: boolean;
+  isQueryRunning: boolean;
+
+  loadQuestion(): Promise<SdkQuestionResult & { originalQuestion?: Question }>;
+  onQuestionChange(question: Question): Promise<void>;
+  onNavigateToNewCard(params: NavigateToNewCardParams): Promise<void>;
+}
+
+export function useLoadQuestion({
+  location,
+  params,
+}: LoadSdkQuestionParams): LoadQuestionHookResult {
+  const dispatch = useDispatch();
+
+  // Keep track of the latest question and query results.
+  // They can be updated from the below actions.
+  const [result, setQuestionResult] = useState<SdkQuestionResult>({});
+  const { question, queryResults } = result;
+
+  const deferredRef = useRef<Deferred>();
+
+  function deferred() {
+    // Cancel the previous query when a new one is started.
+    deferredRef.current?.resolve();
+    deferredRef.current = defer();
+
+    return deferredRef.current;
+  }
```

Comment:
Is this how we are supposed to do request cancellation with deferred? I find these examples in the codebase, but not sure if we are doing it right.